src/ibek/ioc_factory.py

```diff
@@ -0,0 +1,225 @@
+"""
+functions for making Entity models
+"""
+
+from __future__ import annotations
+
+import builtins
+from pathlib import Path
+from typing import Annotated, Any, Dict, List, Literal, Sequence, Tuple, Type, Union
+
+from pydantic import Field, create_model, field_validator
+from pydantic.fields import FieldInfo
+from pydantic_core import PydanticUndefined
+from ruamel.yaml.main import YAML
+
+from ibek.utils import UTILS
+
+from .args import EnumArg, IdArg, ObjectArg
+from .ioc import IOC, Entity, EnumVal, clear_entity_model_ids, get_entity_by_id
+from .support import Definition, Support
+
+
+class IocFactory:
+    """
+    A Class for creating Generic IOC classes and instances
+    """
+
+    def __init__(self) -> None:
+        self._entity_classes: Dict[str, Type[Entity]] = {}
+
+    def ioc_deserialize(
+        self, ioc_instance_yaml: Path, definition_yaml: List[Path]
+    ) -> IOC:
+        """
+        Takes an ioc instance entities file, list of generic ioc definitions files.
+
+        Returns a model of the resulting ioc instance
+        """
+        ioc_model = self.ioc_create_model(definition_yaml)
+
+        # extract the ioc instance yaml into a dict
+        ioc_instance_dict = YAML(typ="safe").load(ioc_instance_yaml)
+
+        if ioc_instance_dict is None or "ioc_name" not in ioc_instance_dict:
+            raise RuntimeError(
+                f"Failed to load a valid ioc config from {ioc_instance_yaml}"
+            )
+
+        # extract the ioc name into UTILS for use in jinja renders
+        name = UTILS.render({}, ioc_instance_dict["ioc_name"])
+        UTILS.set_ioc_name(name)
+        ioc_instance_dict["ioc_name"] = name
+
+        # Create an IOC instance from the instance dict and the model
+        ioc_instance = ioc_model(**ioc_instance_dict)
+        # post processing to insert SubEntity instances
+        self._process_collections(ioc_instance)
+
+        return ioc_instance
+
+    def ioc_create_model(self, definitions: List[Path]) -> Type[IOC]:
+        """
+        Take a list of definitions YAML and create an IOC model from it
+        """
+        entity_models = []
+
+        clear_entity_model_ids()
```

Comment:
Right. Well I moved it into their so I only needed to do it once but it is there to support creating more than one IOC in one session. Only tests care about this because the CLI for ibek only ever does one IOC instance at a time.

Does this line worry you at all?

Reply:
Only that the docstring for the function says "Used for testing", which I perhaps wrongly interpreted as _only_ used for testing.

Reply:
OK I'll fix that then. I would love for the refactor to remove the need for the global list of id->object but probably it won't.

Reply:
as mentioned in another comment there is now only one clear_entity_ids